### PR TITLE
FuriHal: UART refactoring

### PR DESCRIPTION
**Pull request by skotopes**
# What's new

- FuriHal: UART refactoring

# Verification 

- Full set of tests

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: applications/settings/system/system_settings.c [hunk 1]
```diff
@@ -30,6 +30,53 @@ static void log_level_changed(VariableItem* item) {
...
+    "None",
+};
+
+const uint32_t log_device_value[] = {
+    FuriHalRtcLogDeviceUsart,
+    FuriHalRtcLogDeviceLpuart,
+    FuriHalRtcLogDeviceNone};
+
+static void log_device_changed(VariableItem* item) {
+    // SystemSettings* app = variable_item_get_context(item);
```
Comment (hedger) [inline applications/settings/system/system_settings.c]: commented out

File: applications/settings/system/system_settings.c [hunk 2]
```diff
@@ -30,6 +30,53 @@ static void log_level_changed(VariableItem* item) {
...
+    FuriHalRtcLogBaudRate57600,
+    FuriHalRtcLogBaudRate115200,
+    FuriHalRtcLogBaudRate230400,
+    FuriHalRtcLogBaudRate460800,
+    FuriHalRtcLogBaudRate921600,
+    FuriHalRtcLogBaudRate1843200,
+};
+
+static void log_baud_rate_changed(VariableItem* item) {
+    // SystemSettings* app = variable_item_get_context(item);
```
Comment (hedger) [inline applications/settings/system/system_settings.c]: commented out

File: furi/core/log.h [hunk 3]
```diff
@@ -39,12 +39,45 @@ typedef enum {
...
+bool furi_log_remove_handler(FuriLogHandler handler);
+
+/** Transmit data throw log IO callbacks
+ *
+ * @param[in]  data  The data
+ * @param[in]  size  The size
+ */
+void furi_log_tx(const uint8_t* data, size_t size);
+
+/** Transmit data throw log IO callbacks
```
Comment (hedger) [inline furi/core/log.h]: C+P

File: furi/core/log.h [hunk 4]
```diff
@@ -39,12 +39,45 @@ typedef enum {
...
+
+/** Remove log TX callback
+ *
+ * @param[in]  callback  The callback
+ *
+ * @return     true on success, false otherwise
+ */
+bool furi_log_remove_handler(FuriLogHandler handler);
+
+/** Transmit data throw log IO callbacks
```
Comment (hedger) [inline furi/core/log.h]: Through?

File: furi/core/log.c [hunk 5]
```diff
@@ -32,9 +34,75 @@ static const FuriLogLevelDescription FURI_LOG_LEVEL_DESCRIPTIONS[] = {
...
+    while(!FuriLogHandlersList_end_p(it)) {
+        if(memcmp(FuriLogHandlersList_ref(it), &handler, sizeof(FuriLogHandler)) == 0) {
+            ret = false;
+        } else {
+            FuriLogHandlersList_next(it);
+        }
+    }
+
+    if(ret) {
+        FuriLogHandlersList_push_back(furi_log.tx_handlers, handler);
```
Comment (hedger) [inline furi/core/log.c:59]: maybe furi_check that ->callback != nullptr?

File: furi/core/log.c [hunk 6]
```diff
@@ -32,9 +34,75 @@ static const FuriLogLevelDescription FURI_LOG_LEVEL_DESCRIPTIONS[] = {
...
+    bool ret = false;
+
+    furi_check(furi_mutex_acquire(furi_log.mutex, FuriWaitForever) == FuriStatusOk);
+
+    FuriLogHandlersList_it_t it;
+    FuriLogHandlersList_it(it, furi_log.tx_handlers);
+    while(!FuriLogHandlersList_end_p(it)) {
+        if(memcmp(FuriLogHandlersList_ref(it), &handler, sizeof(FuriLogHandler)) == 0) {
+            FuriLogHandlersList_remove(furi_log.tx_handlers, it);
+            ret = true;
```
Comment (hedger) [inline furi/core/log.c:77]: also break out of loop?

File: targets/f18/api_symbols.csv [hunk 7]
```diff
@@ -1,5 +1,5 @@
 entry,status,name,type,params
-Version,+,44.0,,
+Version,+,46.0,,
```
Comment (hedger) [inline targets/f18/api_symbols.csv]: 44->46

File: targets/f7/api_symbols.csv [hunk 8]
```diff
@@ -1,5 +1,5 @@
 entry,status,name,type,params
-Version,+,44.0,,
+Version,+,46.0,,
```
Comment (hedger) [inline targets/f7/api_symbols.csv]: 44->46

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 9]
```diff
@@ -198,47 +207,64 @@ void furi_hal_uart_tx(FuriHalUartId ch, uint8_t* buffer, size_t buffer_size) {
...
     }
 }
 
-void USART1_IRQHandler(void) {
-    if(LL_USART_IsActiveFlag_RXNE_RXFNE(USART1)) {
-        uint8_t data = LL_USART_ReceiveData8(USART1);
-        irq_cb[FuriHalUartIdUSART1](UartIrqEventRXNE, data, irq_ctx[FuriHalUartIdUSART1]);
-    } else if(LL_USART_IsActiveFlag_ORE(USART1)) {
-        LL_USART_ClearFlag_ORE(USART1);
+void furi_hal_serial_set_rx_callback(
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: furi_check args?

File: targets/f7/furi_hal/furi_hal_serial.h [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+ * Real transmission will be completed later. Use
+ * `furi_hal_serial_tx_wait_complete` to wait for completion if you need it.
+ *
+ * @param      handle       Serial handle
+ * @param      buffer       data
+ * @param      buffer_size  data size (in bytes)
+ */
+void furi_hal_serial_tx(FuriHalSerialHandle* handle, const uint8_t* buffer, size_t buffer_size);
+
+/** Wait transmission complete.
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.h]: Wait until transmission is completed

File: furi/core/log.h [hunk 11]
```diff
@@ -76,18 +109,6 @@ void furi_log_set_level(FuriLogLevel level);
  */
 FuriLogLevel furi_log_get_level();
```
Comment (hedger) [inline furi/core/log.h]: (void);

File: furi/core/log.h [hunk 12]
```diff
@@ -39,12 +39,45 @@ typedef enum {
...
-typedef uint32_t (*FuriLogTimestamp)(void);
+typedef void (*FuriLogHandlerCallback)(const uint8_t* data, size_t size, void* context);
+
+typedef struct {
+    FuriLogHandlerCallback callback;
+    void* context;
+} FuriLogHandler;
 
 /** Initialize logging */
 void furi_log_init();
```
Comment (hedger) [inline furi/core/log.h]: (void);

File: targets/f7/furi_hal/furi_hal_rtc.h [hunk 13]
```diff
@@ -118,6 +136,30 @@ void furi_hal_rtc_set_log_level(uint8_t level);
...
+ *
+ * @param[in]  baud_rate  The baud rate
+ */
+void furi_hal_rtc_set_log_baud_rate(FuriHalRtcLogBaudRate baud_rate);
+
+/** Get logging baud rate
+ *
+ * @return     The furi hal rtc log baud rate.
+ */
+FuriHalRtcLogBaudRate furi_hal_rtc_get_log_baud_rate();
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_rtc.h]: (void);

File: targets/f7/furi_hal/furi_hal_rtc.h [hunk 14]
```diff
@@ -118,6 +136,30 @@ void furi_hal_rtc_set_log_level(uint8_t level);
...
+ *
+ * @param[in]  device  The device
+ */
+void furi_hal_rtc_set_log_device(FuriHalRtcLogDevice device);
+
+/** Get logging device
+ *
+ * @return     The furi hal rtc log device.
+ */
+FuriHalRtcLogDevice furi_hal_rtc_get_log_device();
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_rtc.h]: (void);

File: targets/f7/furi_hal/furi_hal_rtc.h [hunk 15]
```diff
@@ -118,6 +136,30 @@ void furi_hal_rtc_set_log_level(uint8_t level);
  */
 uint8_t furi_hal_rtc_get_log_level();
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_rtc.h]: (void);

File: targets/f7/furi_hal/furi_hal_rtc.h [hunk 16]
```diff
@@ -64,21 +64,39 @@ typedef enum {
...
+    FuriHalRtcLogBaudRate38400 = 0x2, /**< 38400 baud */
+    FuriHalRtcLogBaudRate57600 = 0x3, /**< 57600 baud */
+    FuriHalRtcLogBaudRate115200 = 0x4, /**< 115200 baud */
+    FuriHalRtcLogBaudRate460800 = 0x5, /**< 460800 baud */
+    FuriHalRtcLogBaudRate921600 = 0x6, /**< 921600 baud */
+    FuriHalRtcLogBaudRate1843200 = 0x7, /**< 1843200 baud */
+} FuriHalRtcLogBaudRate;
+
 /** Early initialization */
 void furi_hal_rtc_init_early();
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_rtc.h]: (void);

**Comment by 0xchocolate** [comment]
Just want to make sure I'm reading this right and understand the impact - this will break all apps that use uart until they migrate to the new API?

And once this merges, I have to decide if I want to support either dev or release SDK, but not both for a while?

**Comment by Tetr0N** [comment]
> **Compiled f7 firmware for commit `e6b6af02`:**
> - [📦 Update package](https://update.flipperzero.one/builds/firmware/aku/uart/flipper-z-f7-update-aku_uart-15112023-e6b6af02.tgz)
> - [📥 DFU file](https://update.flipperzero.one/builds/firmware/aku/uart/flipper-z-f7-full-aku_uart-15112023-e6b6af02.dfu)
> - [☁️ Web/App updater](https://lab.flipper.net/?url=https://update.flipperzero.one/builds/firmware/aku/uart/flipper-z-f7-update-aku_uart-15112023-e6b6af02.tgz&channel=aku/uart&version=e6b6af02)

**Review by luminus82 — COMMENTED** [review]
Tom

**Comment by skotopes** [comment]
> Just want to make sure I'm reading this right and understand the impact - this will break all apps that use uart until they migrate to the new API?
> 
> And once this merges, I have to decide if I want to support either dev or release SDK, but not both for a while?

Yes.

We've designed app catalog flow in a way that it's not going to impact users, all you need is to submit app in advance.

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 17]
```diff
@@ -0,0 +1,797 @@
...
+    if(USART1->ISR & USART_ISR_FE) {
+        USART1->ICR = USART_ICR_FECF;
+        event |= FuriHalSerialRxEventFrameError;
+    }
+    if(USART1->ISR & USART_ISR_PE) {
+        USART1->ICR = USART_ICR_PECF;
+        event |= FuriHalSerialRxEventFrameError;
+    }
+
+    if(furi_hal_serial[FuriHalSerialIdUsart].buffer_rx_ptr == NULL) {
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:73]: minor speedup: 
`const ...* usart_serial = &furi_hal_serial[FuriHalSerialIdUsart];`
then
`usart_serial->buffer_rx_ptr` etc etc

reason: does not index the array on each access, better readability

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 18]
```diff
@@ -0,0 +1,797 @@
...
+                furi_hal_serial[FuriHalSerialIdUsart].context);
+        }
+    }
+}
+
+static void furi_hal_serial_usart_dma_rx_isr() {
+#if LL_DMA_CHANNEL_6 == LL_DMA_CHANNEL_6
+    if(LL_DMA_IsActiveFlag_HT6(DMA1)) {
+        LL_DMA_ClearFlag_HT6(DMA1);
+        furi_hal_serial[FuriHalSerialIdUsart].buffer_rx_index_write =
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:96]: same here - cache pointer to entry

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 19]
```diff
@@ -0,0 +1,797 @@
...
+        }
+    }
+#else
+#error Update this code. Would you kindly?
+#endif
+}
+
+static void furi_hal_serial_usart_init_dma_rx(void) {
+    /* USART1_RX_DMA Init */
+    furi_check(furi_hal_serial[FuriHalSerialIdUsart].buffer_rx_ptr == NULL);
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:134]: same here

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 20]
```diff
@@ -0,0 +1,797 @@
...
+        LL_DMA_CHANNEL_6,
+        LL_DMA_DIRECTION_PERIPH_TO_MEMORY | LL_DMA_MODE_CIRCULAR | LL_DMA_PERIPH_NOINCREMENT |
+            LL_DMA_MEMORY_INCREMENT | LL_DMA_PDATAALIGN_BYTE | LL_DMA_MDATAALIGN_BYTE |
+            LL_DMA_PRIORITY_HIGH);
+    LL_DMA_SetDataLength(DMA1, LL_DMA_CHANNEL_6, FURI_HAL_SERIAL_DMA_BUFFER_SIZE);
+    LL_DMA_SetPeriphRequest(DMA1, LL_DMA_CHANNEL_6, LL_DMAMUX_REQ_USART1_RX);
+
+    furi_hal_interrupt_set_isr(FuriHalInterruptIdDma1Ch6, furi_hal_serial_usart_dma_rx_isr, NULL);
+
+#if LL_DMA_CHANNEL_6 == LL_DMA_CHANNEL_6
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: always true

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 21]
```diff
@@ -0,0 +1,797 @@
...
+                furi_hal_serial[FuriHalSerialIdUsart].handle,
+                event,
+                furi_hal_serial_dma_bytes_available(FuriHalSerialIdUsart),
+                furi_hal_serial[FuriHalSerialIdUsart].context);
+        }
+    }
+}
+
+static void furi_hal_serial_usart_dma_rx_isr() {
+#if LL_DMA_CHANNEL_6 == LL_DMA_CHANNEL_6
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: always true

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 22]
```diff
@@ -0,0 +1,797 @@
...
+        LL_DMA_ClearFlag_HT7(DMA1);
+
+        LL_DMA_DeInit(DMA1, LL_DMA_CHANNEL_6);
+        furi_hal_interrupt_set_isr(FuriHalInterruptIdDma1Ch6, NULL, NULL);
+        free(furi_hal_serial[FuriHalSerialIdUsart].buffer_rx_ptr);
+        furi_hal_serial[FuriHalSerialIdUsart].buffer_rx_ptr = NULL;
+    }
+}
+
+static void furi_hal_serial_uasrt_init(FuriHalSerialHandle* handle, uint32_t baud) {
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: uarst - typo

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 23]
```diff
@@ -0,0 +1,797 @@
...
+    if(LPUART1->ISR & USART_ISR_FE) {
+        LPUART1->ICR = USART_ICR_FECF;
+        event |= FuriHalSerialRxEventFrameError;
+    }
+    if(LPUART1->ISR & USART_ISR_PE) {
+        LPUART1->ICR = USART_ICR_PECF;
+        event |= FuriHalSerialRxEventFrameError;
+    }
+
+    if(furi_hal_serial[FuriHalSerialIdLpuart].buffer_rx_ptr == NULL) {
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:274]: maybe cache here too

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 24]
```diff
@@ -0,0 +1,797 @@
...
+                furi_hal_serial[FuriHalSerialIdLpuart].handle,
+                event,
+                furi_hal_serial_dma_bytes_available(FuriHalSerialIdLpuart),
+                furi_hal_serial[FuriHalSerialIdLpuart].context);
+        }
+    }
+}
+
+static void furi_hal_serial_lpuart_dma_rx_isr() {
+#if LL_DMA_CHANNEL_7 == LL_DMA_CHANNEL_7
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: always true

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 25]
```diff
@@ -0,0 +1,797 @@
...
+    if(LL_DMA_IsActiveFlag_HT7(DMA1)) {
+        LL_DMA_ClearFlag_HT7(DMA1);
+        furi_hal_serial[FuriHalSerialIdLpuart].buffer_rx_index_write =
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:297]: maybe cache here too

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 26]
```diff
@@ -0,0 +1,797 @@
...
+        LL_DMA_CHANNEL_7,
+        LL_DMA_DIRECTION_PERIPH_TO_MEMORY | LL_DMA_MODE_CIRCULAR | LL_DMA_PERIPH_NOINCREMENT |
+            LL_DMA_MEMORY_INCREMENT | LL_DMA_PDATAALIGN_BYTE | LL_DMA_MDATAALIGN_BYTE |
+            LL_DMA_PRIORITY_HIGH);
+    LL_DMA_SetDataLength(DMA1, LL_DMA_CHANNEL_7, FURI_HAL_SERIAL_DMA_BUFFER_SIZE);
+    LL_DMA_SetPeriphRequest(DMA1, LL_DMA_CHANNEL_7, LL_DMAMUX_REQ_LPUART1_RX);
+
+    furi_hal_interrupt_set_isr(FuriHalInterruptIdDma1Ch7, furi_hal_serial_lpuart_dma_rx_isr, NULL);
+
+#if LL_DMA_CHANNEL_7 == LL_DMA_CHANNEL_7
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: always true

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 27]
```diff
@@ -0,0 +1,797 @@
...
+        }
+    }
+#else
+#error Update this code. Would you kindly?
+#endif
+}
+
+static void furi_hal_serial_lpuart_init_dma_rx(void) {
+    /* LPUART1_RX_DMA Init */
+    furi_check(furi_hal_serial[FuriHalSerialIdLpuart].buffer_rx_ptr == NULL);
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:335]: maybe cache here too

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 28]
```diff
@@ -0,0 +1,797 @@
...
+        if(furi_hal_bus_is_enabled(FuriHalBusUSART1)) {
+            furi_hal_bus_disable(FuriHalBusUSART1);
+        }
+        if(LL_USART_IsEnabled(USART1)) {
+            LL_USART_Disable(USART1);
+        }
+        furi_hal_serial_usart_deinit_dma_rx();
+        furi_hal_gpio_init(&gpio_usart_tx, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_gpio_init(&gpio_usart_rx, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_serial[handle->id].enabled = false;
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: can be moved outside "if", like in function below

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 29]
```diff
@@ -0,0 +1,797 @@
...
+            LL_LPUART_Enable(LPUART1);
+        }
+    }
+}
+
+void furi_hal_serial_deinit(FuriHalSerialHandle* handle) {
+    furi_check(handle);
+    furi_hal_serial_async_rx_configure(handle, NULL, NULL);
+    if(handle->id == FuriHalSerialIdUsart) {
+        if(furi_hal_bus_is_enabled(FuriHalBusUSART1)) {
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:554]: deinit bus later?

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 30]
```diff
@@ -0,0 +1,797 @@
...
+        furi_hal_serial_lpuart_deinit_dma_rx();
+        furi_hal_gpio_init(&gpio_ext_pc0, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_gpio_init(&gpio_ext_pc1, GpioModeAnalog, GpioPullNo, GpioSpeedLow);
+        furi_hal_serial[handle->id].enabled = false;
+    }
+}
+
+void furi_hal_serial_suspend(FuriHalSerialHandle* handle) {
+    furi_check(handle);
+    if(handle->id == FuriHalSerialIdLpuart && LL_LPUART_IsEnabled(LPUART1)) {
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:581]: maybe also check `[].enabled` like below?

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 31]
```diff
@@ -0,0 +1,797 @@
...
+            LL_USART_Enable(USART1);
+        }
+        furi_hal_serial[handle->id].enabled = true;
+    }
+}
+
+void furi_hal_serial_tx(FuriHalSerialHandle* handle, const uint8_t* buffer, size_t buffer_size) {
+    furi_check(handle);
+    if(handle->id == FuriHalSerialIdUsart) {
+        if(LL_USART_IsEnabled(USART1) == 0) return;
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:604]: inconsistent enabled checks - in some places it's LL_*, sometimes [].enabled

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 32]
```diff
@@ -0,0 +1,797 @@
...
+            buffer++;
+            buffer_size--;
+        }
+    }
+}
+
+void furi_hal_serial_tx_wait_complete(FuriHalSerialHandle* handle) {
+    furi_check(handle);
+    if(handle->id == FuriHalSerialIdUsart) {
+        if(LL_USART_IsEnabled(USART1) == 0) return;
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:633]: same here

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 33]
```diff
@@ -0,0 +1,797 @@
...
+    if(handle->id == FuriHalSerialIdUsart) {
+        return LL_USART_ReceiveData8(USART1);
+    }
+    return LL_LPUART_ReceiveData8(LPUART1);
+}
+
+static size_t furi_hal_serial_dma_bytes_available(FuriHalSerialId ch) {
+    size_t index_dma = 0;
+    if(ch == FuriHalSerialIdUsart) {
+        index_dma = LL_DMA_GetDataLength(DMA1, LL_DMA_CHANNEL_6);
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: move channel to separate `#define`?

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 34]
```diff
@@ -0,0 +1,797 @@
...
+    } else if(ch == FuriHalSerialIdLpuart) {
+        index_dma = LL_DMA_GetDataLength(DMA1, LL_DMA_CHANNEL_7);
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: same

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 35]
```diff
@@ -0,0 +1,797 @@
...
+    } else {
+        furi_crash("Invalid programming");
+    }
+
+    furi_hal_serial[ch].buffer_rx_index_write = FURI_HAL_SERIAL_DMA_BUFFER_SIZE - index_dma;
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: cache pointer?

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 36]
```diff
@@ -0,0 +1,797 @@
...
+    } else {
+        return FURI_HAL_SERIAL_DMA_BUFFER_SIZE - furi_hal_serial[ch].buffer_rx_index_read +
+               furi_hal_serial[ch].buffer_rx_index_write;
+    }
+}
+
+static uint8_t furi_hal_serial_dma_rx_read_byte(FuriHalSerialHandle* handle) {
+    uint8_t data = 0;
+    data =
+        furi_hal_serial[handle->id].buffer_rx_ptr[furi_hal_serial[handle->id].buffer_rx_index_read];
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:762]: same

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 37]
```diff
@@ -0,0 +1,797 @@
...
+static void furi_hal_serial_dma_configure(
+    FuriHalSerialHandle* handle,
+    FuriHalSerialDmaRxCallback callback,
+    void* context) {
+    furi_check(handle);
+    if(handle->id == FuriHalSerialIdUsart) {
+        if(callback == NULL) {
+            furi_hal_interrupt_set_isr(FuriHalInterruptIdUart1, NULL, NULL);
+            //deinit DMA
+            furi_hal_serial_usart_init_dma_rx();
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c]: typo? init, deinit in comment

File: targets/f7/furi_hal/furi_hal_serial.c [hunk 38]
```diff
@@ -0,0 +1,797 @@
...
+            //disable IRQ RX byte
+            LL_USART_DisableIT_RXNE_RXFNE(USART1);
+        } else {
+            //deinit DMA
+            furi_hal_serial_usart_init_dma_rx();
```
Comment (hedger) [inline targets/f7/furi_hal/furi_hal_serial.c:792]: deinit?..

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 35 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: No need, must be refactored later.

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 36 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:762]: same

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 33 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: will be refactored later

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 34 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: same

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 31 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:604]: yes, whole thing must be refactored, but not in this iteration

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 28 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: done

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 29 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:554]: must be refactored, but not now

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 27 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:335]: will be refactored

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 26 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: fixed

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 25 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:297]: will be refactored

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 24 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: fixed

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 23 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:274]: will be refactored

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 22 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: fixed

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 21 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: fixed

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 20 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c]: fixed

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 19 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:134]: will be refactored

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 18 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:96]: will be refactored

File: targets/f7/furi_hal/furi_hal_serial.c (hunk 17 above)
Comment (skotopes) [reply targets/f7/furi_hal/furi_hal_serial.c:73]: will be refactored

**Review by hedger — APPROVED** [review]

**Comment by skotopes** [comment]
Porting to new API example: https://github.com/skotopes/flipperzero_mhz19_uart/commit/9612b6fcac2db9f819db0aad3683a95db5ff28ea

**Comment by andreww24** [comment]
Hello! It is possible to add ext function for uart initialization?
With additional parameters, like parity, number stop bit, inverse level of channels( RXINV and TXINV)?
Inverse level is frequently needed for some 1-wire uart physical impementation.
Thank you.

**Pull request merged**